README.md
# halli
[![License: MIT](https://img.shields.io/badge/License-MIT-brightgreen.svg)](https://github.com/clok/halli/blob/master/LICENSE)
[![Coverage Status](https://coveralls.io/repos/github/clok/halli/badge.svg)](https://coveralls.io/github/clok/halli)

Color gradient generator and picker

[<img src="https://clok.sh/assets/img/color-select-sample.png">](https://clok.sh/demos/halli)

## Why?

Many moons ago I had a need to efficiently generate colors for [particle effects in a game engine](https://clokwork.net/articles/particle-generation-in-impactjs/).
It was a fun problem to work on and resulted in writing [some fun code](https://clokwork.net/articles/simplifying-color-selection/) to
dynamically generate color gradients that can be used. I published an [ImpactJS plugin](https://github.com/clok/impactjs-color-picker) as well. 

Recently, I have found the need for generating colors. So, here we are.

## Installation

As a module
```text
$ npm install --save halli
```

As a script tag
```html
<script type="text/javascript" src="https://clok.sh/assets/js/halli.v0.3.1.min.js" />
<script type="text/javascript">
  // Initialize Halli object
  var picker = new Halli;

  // Generate and draw fade between White and Black with 400 steps
  var fade = picker.genHexArray(0xFFFFFF, 0x000000, 400);

  // Generate the Matlab Jet colormap with 80 steps between each color
  var jet = picker.genMultiHexArray([0x0000FF, 0x00FFFF, 0x00FF00, 0xFFFF00, 0xFF0000, 0x000000], 80);

  // Generate the Matlab HSV colormap with 70 steps between each color
  var hsv = picker.genMultiHexArray([0xFF0000, 0xFFFF00, 0x00FF00, 0x00FFFF, 0x0000FF, 0xFF00FF, 0x000000], 70);
</script>
```

## Demo

See [demo](https://clok.sh/demos/halli)


## Usage

> See [`Halli` class documentation](./docs/classes/Halli.md)

```typescript
import { Halli } from 'halli'

// Initialize ColorPicker object
const picker = new Halli()

// Generate and draw fade between White and Black with 400 steps
const fade = picker.genHexArray(0xffffff, 0x000000, 400)

/**
 * Generate the Matlab Jet colormap with 80 steps between each color
 * then render
 */
const jet = picker.genMultiHexArray(
  [0x0000ff, 0x00ffff, 0x00ff00, 0xffff00, 0xff0000, 0x000000],
  80,
)

/**
 * Generate the Matlab HSV colormap with 70 steps between each color
 * then render
 */
const hsv = picker.genMultiHexArray(
  [0xff0000, 0xffff00, 0x00ff00, 0x00ffff, 0x0000ff, 0xff00ff, 0x000000],
  70,
)

/**
 * Regenerate the Matlab HSV colormap in a higher resolution with
 * 27,000 steps between each color.
 * Use the createImageData method to quickly render the colormap
 * with each entry in the array represented by a single pixel.
 */
const pix = picker.genMultiHexArray(
  [0xff0000, 0xffff00, 0x00ff00, 0x00ffff, 0x0000ff, 0xff00ff, 0x000000],
  27000,
)
```

## Development

1. Fork the [clok/halli](https://github.com/clok/halli) repo
2. Branch & Code
3. Run linters :broom: `npm run lint`
4. Commit with a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/)
5. Open a PR

### Local Demo

1. Clone the repo
2. `npm install`
3. `npm build`
4. `open ./demo.html`

#### Why is it named `halli`?

According to [Google translate](https://www.google.com/search?q=gradient+in+icelandic) "gradient" in English translated to Icelandic results in "halli".


CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]


<a name="v1.0.0"></a>
## [v1.0.0] - 2022-06-10
### Chore
- update readme
- **deps:** update all non-major dependencies ([#22](https://github.com/clok/halli/issues/22))
- **deps:** update all non-major dependencies ([#21](https://github.com/clok/halli/issues/21))
- **docs:** updating docs for version v1.0.0

### Feat
- **css color:** support CSS style 6 digit hex color codes ([#23](https://github.com/clok/halli/issues/23))


<a name="v0.3.1"></a>
## [v0.3.1] - 2022-06-03
### Chore
- **docs:** updating docs for version v0.3.1
- **release:** updating CHANGELOG for v0.3.1

### Fix
- patch relase process


<a name="v0.3.0"></a>
## [v0.3.0] - 2022-06-03
### Chore
- **deps:** update all non-major dependencies ([#15](https://github.com/clok/halli/issues/15))
- **deps:** update all non-major dependencies ([#10](https://github.com/clok/halli/issues/10))
- **deps:** update jest monorepo to v28 ([#16](https://github.com/clok/halli/issues/16))
- **deps:** update dependency husky to v8 ([#17](https://github.com/clok/halli/issues/17))
- **deps:** pin dependency uglify-js to 3.15.5 ([#20](https://github.com/clok/halli/issues/20))
- **deps:** update goreleaser/goreleaser-action action to v3 ([#18](https://github.com/clok/halli/issues/18))
- **deps:** update dependency eslint-plugin-jest to v26 ([#11](https://github.com/clok/halli/issues/11))
- **deps:** update actions/setup-node action to v3 ([#12](https://github.com/clok/halli/issues/12))
- **deps:** update actions/setup-go action to v3 ([#13](https://github.com/clok/halli/issues/13))
- **deps:** update actions/checkout action to v3 ([#14](https://github.com/clok/halli/issues/14))
- **docs:** updating docs for version v0.3.0
- **release:** updating CHANGELOG for v0.3.0

### Feat
- add support for minified version that can be loaded in HTML script tag ([#19](https://github.com/clok/halli/issues/19))


<a name="v0.2.1"></a>
## [v0.2.1] - 2022-01-06
### Chore
- update readme links
- update README
- **deps:** update all non-major dependencies ([#9](https://github.com/clok/halli/issues/9))
- **docs:** updating docs for version v0.2.1
- **release:** updating CHANGELOG for v0.2.1
- **test:** update test cases


<a name="v0.2.0"></a>
## [v0.2.0] - 2021-12-31
### Chore
- **docs:** updating docs for version v0.2.0
- **release:** updating CHANGELOG for v0.2.0

### Fix
- **genMultiHexArray:** correct color delta math


<a name="v0.1.2"></a>
## [v0.1.2] - 2021-12-30
### Chore
- create release of tag push
- **deps:** update all non-major dependencies to v5.8.0 ([#7](https://github.com/clok/halli/issues/7))
- **deps:** update all non-major dependencies ([#6](https://github.com/clok/halli/issues/6))
- **deps:** update all non-major dependencies ([#8](https://github.com/clok/halli/issues/8))
- **docs:** updating docs for version v0.1.1
- **docs:** updating docs for version v0.1.2
- **release:** updating CHANGELOG for v0.1.1
- **release:** updating CHANGELOG for v0.1.2


<a name="v0.1.1"></a>
## [v0.1.1] - 2021-12-17
### Chore
- add ci tests
- regenerate lock file
- **deps:** update actions/checkout action to v2 ([#4](https://github.com/clok/halli/issues/4))
- **deps:** update node version on github workflow
- **deps:** update actions/setup-node action to v2 ([#5](https://github.com/clok/halli/issues/5))
- **deps:** update node.js to v16 ([#2](https://github.com/clok/halli/issues/2))
- **deps:** add renovate.json ([#1](https://github.com/clok/halli/issues/1))
- **docs:** updating docs for version v0.1.1
- **docs:** updating docs for version v0.1.0
- **release:** updating CHANGELOG for v0.1.1
- **release:** updating CHANGELOG for v0.1.0


<a name="v0.1.0"></a>
## v0.1.0 - 2021-12-17
### Chore
- add initial docs
- update readme
- add husky
- add release helper script
- generate docs
- **changelog:** generate initial changelog
- **docs:** updating docs for version v0.1.0
- **release:** updating CHANGELOG for v0.1.0

### Feat
- improved documentation and cleaned up code
- **ts:** initial port to typescript


[Unreleased]: https://github.com/clok/halli/compare/v1.0.0...HEAD
[v1.0.0]: https://github.com/clok/halli/compare/v0.3.1...v1.0.0
[v0.3.1]: https://github.com/clok/halli/compare/v0.3.0...v0.3.1
[v0.3.0]: https://github.com/clok/halli/compare/v0.2.1...v0.3.0
[v0.2.1]: https://github.com/clok/halli/compare/v0.2.0...v0.2.1
[v0.2.0]: https://github.com/clok/halli/compare/v0.1.2...v0.2.0
[v0.1.2]: https://github.com/clok/halli/compare/v0.1.1...v0.1.2
[v0.1.1]: https://github.com/clok/halli/compare/v0.1.0...v0.1.1


src/index.spec.ts
import {Halli} from './index'

describe('Halli', () => {
  describe('pickHex', () => {
    it('consistently returns the same value for a ratio in range', () => {
      const picker = new Halli()
      expect(picker.pickHex(0x000000, 0xffffff, 0.5)).toBe('#7f7f7f')
    })

    it('consistently returns the same value for a ratio in range [ratio > 1]', () => {
      const picker = new Halli()
      expect(picker.pickHex(0x000000, 0xffffff, 1.5)).toBe('#ffffff')
    })

    it('pick correct length color [edge case]', () => {
      const picker = new Halli()
      expect(picker.pickHex(0xffffff, 0xff0000, 0)).toBe('#ffffff')
      expect(picker.pickHex(0xffffff, 0xff0000, 1)).toBe('#ff0000')
      expect(picker.pickHex(0xffffff, 0xff0000, 0.5)).toBe('#ff7f7f')
      expect(picker.pickHex(0xffffff, 0xff0000, Math.random())).toHaveLength(7)
      expect(picker.pickHex(16777215, 16711680, 0.5)).toBe('#ff7f7f')
    })

    it('consistently returns the same value for a ratio in range [hex string]', () => {
      const picker = new Halli()
      expect(picker.pickHex('#000000', '#ffffff', 0.5)).toBe('#7f7f7f')
    })

    it('consistently returns the same value for a ratio in range [ratio > 1 hex string]', () => {
      const picker = new Halli()
      expect(picker.pickHex('#000000', '#ffffff', 1.5)).toBe('#ffffff')
    })

    it('pick correct length color [edge case hex string]', () => {
      const picker = new Halli()
      expect(picker.pickHex('#ffffff', '#ff0000', 0)).toBe('#ffffff')
      expect(picker.pickHex('#ffffff', '#ff0000', 1)).toBe('#ff0000')
      expect(picker.pickHex('#ffffff', '#ff0000', 0.5)).toBe('#ff7f7f')
      expect(picker.pickHex('#ffffff', '#ff0000', Math.random())).toHaveLength(
        7,
      )
      expect(picker.pickHex('16777215', '16711680', 0.5)).toBe('#ff7f7f')
    })
  })

  describe('genHexArray', () => {
    it('consistently returns the same value', () => {
      const picker = new Halli()
      const check = picker.genHexArray(0x00000, 0xffffff, 10)
      expect(check.length).toBe(12)
      expect(check).toEqual([
        '#000000',
        '#171717',
        '#2e2e2e',
        '#454545',
        '#5c5c5c',
        '#737373',
        '#8b8b8b',
        '#a2a2a2',
        '#b9b9b9',
        '#d0d0d0',
        '#e7e7e7',
        '#ffffff',
      ])
    })

    it('consistently returns the same value [hex string]', () => {
      const picker = new Halli()
      const check = picker.genHexArray('#000000', '#ffffff', 10)
      expect(check.length).toBe(12)
      expect(check).toEqual([
        '#000000',
        '#171717',
        '#2e2e2e',
        '#454545',
        '#5c5c5c',
        '#737373',
        '#8b8b8b',
        '#a2a2a2',
        '#b9b9b9',
        '#d0d0d0',
        '#e7e7e7',
        '#ffffff',
      ])
    })
  })

  describe('genMultiHexArray', () => {
    it('consistently returns the same value', () => {
      const picker = new Halli()
      const check = picker.genMultiHexArray([0x00ff00, 0xffff00, 0xff0000], 3)
      expect(check.length).toBe(7)
      expect(check).toEqual([
        '#00ff00',
        '#55ff00',
        '#aaff00',
        '#ffff00',
        '#ffaa00',
        '#ff5500',
        '#ff0000',
      ])
    })
  })

  describe('hexToRGBA', () => {
    it('consistently returns the same value', () => {
      let check = Halli.hexToRGBA('#ffffff')
      expect(check.r).toEqual(255)
      expect(check.g).toEqual(255)
      expect(check.b).toEqual(255)
      expect(check.a).toEqual(255)

      check = Halli.hexToRGBA('ffffff')
      expect(check.r).toEqual(255)
      expect(check.g).toEqual(255)
      expect(check.b).toEqual(255)
      expect(check.a).toEqual(255)
    })

    it('consistently returns the same value with alpha', () => {
      let check = Halli.hexToRGBA('#ffffff17')
      expect(check.r).toEqual(255)
      expect(check.g).toEqual(255)
      expect(check.b).toEqual(255)
      expect(check.a).toEqual(23)

      check = Halli.hexToRGBA('ffffff17')
      expect(check.r).toEqual(255)
      expect(check.g).toEqual(255)
      expect(check.b).toEqual(255)
      expect(check.a).toEqual(23)
    })
  })

  describe('hexToRGBAstr', () => {
    it('consistently returns the same value', () => {
      expect(Halli.hexToRGBAstr('#ffffff')).toBe('rgba(255, 255, 255, 255)')
      expect(Halli.hexToRGBAstr('#ae1f8c')).toBe('rgba(174, 31, 140, 255)')
    })

    it('consistently returns the same value with alpha', () => {
      expect(Halli.hexToRGBAstr('#ffffffff')).toBe('rgba(255, 255, 255, 255)')
      expect(Halli.hexToRGBAstr('#ae1f8cff')).toBe('rgba(174, 31, 140, 255)')
      expect(Halli.hexToRGBAstr('#ffffffa4')).toBe('rgba(255, 255, 255, 164)')
      expect(Halli.hexToRGBAstr('#ae1f8c3c')).toBe('rgba(174, 31, 140, 60)')
    })
  })

  describe('convertToColorString', () => {
    it('consistently returns the same value', () => {
      expect(Halli.convertToColorString(0xffffff)).toBe('#ffffff')
      expect(Halli.convertToColorString(0xab12f8)).toBe('#ab12f8')
    })
  })

  describe('convertToNumber', () => {
    class Tester extends Halli {
      static test_convertToNumber(input: string | number): number {
        return Tester.convertToNumber(input)
      }
    }

    it('consistently returns the same value', () => {
      expect(Tester.test_convertToNumber(0xffffff)).toEqual(16777215)
      expect(Tester.test_convertToNumber(0x000000)).toEqual(0)
      expect(Tester.test_convertToNumber(0)).toEqual(0)
      expect(Tester.test_convertToNumber(123456)).toEqual(123456)
      expect(Tester.test_convertToNumber(1234567)).toEqual(1234567)
      expect(Tester.test_convertToNumber(123)).toEqual(123)
      expect(Tester.test_convertToNumber('#ffffff')).toEqual(16777215)
      expect(Tester.test_convertToNumber('#000000')).toEqual(0)
      expect(Tester.test_convertToNumber('0')).toEqual(0)
      expect(Tester.test_convertToNumber('#123456')).toEqual(1193046)
      expect(Tester.test_convertToNumber('1234567')).toEqual(1234567)
      expect(Tester.test_convertToNumber('123')).toEqual(291)
      expect(Tester.test_convertToNumber('#ffffff')).toEqual(16777215)
      expect(Tester.test_convertToNumber('#000000')).toEqual(0)
      expect(Tester.test_convertToNumber('#0')).toEqual(0)
      expect(Tester.test_convertToNumber('#123456')).toEqual(1193046)
      expect(Tester.test_convertToNumber('#123')).toEqual(291)
      expect(Tester.test_convertToNumber('#000123')).toEqual(291)
    })
  })
})


release.sh
#!/bin/bash

set +e

#
# Set Colors
#

bold="\e[1m"
dim="\e[2m"
underline="\e[4m"
blink="\e[5m"
reset="\e[0m"
red="\e[31m"
green="\e[32m"
blue="\e[34m"

#
# Common Output Styles
#

h1() {
  printf "\n${bold}${underline}%s${reset}\n" "$(echo "$@" | sed '/./,$!d')"
}
h2() {
  printf "\n${bold}%s${reset}\n" "$(echo "$@" | sed '/./,$!d')"
}
info() {
  printf "${dim}➜ %s${reset}\n" "$(echo "$@" | sed '/./,$!d')"
}
success() {
  printf "${green}✔ %s${reset}\n" "$(echo "$@" | sed '/./,$!d')"
}
error() {
  printf "${red}${bold}✖ %s${reset}\n" "$(echo "$@" | sed '/./,$!d')"
}
warnError() {
  printf "${red}✖ %s${reset}\n" "$(echo "$@" | sed '/./,$!d')"
}
warnNotice() {
  printf "${blue}✖ %s${reset}\n" "$(echo "$@" | sed '/./,$!d')"
}
note() {
  printf "\n${bold}${blue}Note:${reset} ${blue}%s${reset}\n" "$(echo "$@" | sed '/./,$!d')"
}

typeExists() {
  if [ $(type -P $1) ]; then
    return 0
  fi
  return 1
}

if ! typeExists "git-chglog"; then
  error "git-chglog is not installed"
  note "To install run: go get -u github.com/git-chglog/git-chglog/cmd/git-chglog"
  exit 1
fi

VERSION="v$(grep '\"version\":' package.json | grep -v git-chglog | grep -v release.sh | awk -F '\"' '{print $4}')"

if [ "x${VERSION}x" = "xx" ]; then
  error "Must supply version number as first argument"
  exit 1
fi

h1 "Preparing release of $VERSION"

h2 "Updating distribution"
npm run build
VERSION=${VERSION} npm run minify:release

h2 "Updating docs"
npm run build:docs
if [[ "$(git status -s docs/** 2>/dev/null | wc -l)" == "0" ]]; then
  note "No changes to docs"
else
  note "Committing changes to docs"
  git add docs/**
  git commit -m "chore(docs): updating docs for version $VERSION"
fi

h2 "Updating CHANGELOG.md"
git-chglog --next-tag $VERSION -o CHANGELOG.md && git add CHANGELOG.md
git commit -m "chore(release): updating CHANGELOG for $VERSION"

note "Commands to run:"
note "1. git push origin main"
note "2. git push origin $VERSION"
note "3. npm publish"

success "Done!"


src/min.ts
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck
;(function () {
  // CommonJS module
  if (typeof exports !== 'undefined') {
    if (typeof module !== 'undefined' && module.exports) {
      exports = module.exports = Halli
    }
    exports.Halli = Halli
  }

  // Register as an anonymous AMD module
  if (typeof define === 'function' && define.amd) {
    define([], function () {
      return Halli
    })
  }

  // if there is a importsScrips object define chance for worker
  // allows worker to use full Halli functionality with seed
  if (typeof importScripts !== 'undefined') {
    halli = new Halli()
    self.Halli = Halli
  }

  // If there is a window object, that at least has a document property,
  // instantiate and define chance on the window
  if (typeof window === 'object' && typeof window.document === 'object') {
    window.Halli = Halli
    window.halli = new Halli()
  }
})()


src/halli.ts
/**
 * IRGB rgba representation of a color
 */
export type IRGB = {
  readonly r: number
  readonly g: number
  readonly b: number
  readonly a: number
}

/**
 * IHexColorString is a standard CSS HTML color code. Example: '#AB12F8'
 */
export type IHexColorString = string

/**
 * IHexColorNumber is a hexadecimal number representation of a color. Example: 0xab12f8
 */
export type IHexColorNumber = number

export type IRGBACSSString = `rgba(${number}, ${number}, ${number}, ${number})`

const hexCheck = new RegExp('^#?[0123456789abcdefABCDEF]{1,6}$')

export class Halli {
  colors: IHexColorString[]
  triColors: IHexColorString[]

  /**
   * Initialize object with a default 10 color array fading from
   * White to Red and with a Tri-Color array from Green to Yellow
   * to Red gradient with 10 steps between each source color.
   */
  constructor() {
    // Generate default Colors Array with White to Red in 10 steps
    this.colors = this.genHexArray(0xffffff, 0xff0000, 10)

    // Generate default triColors Array with Green to Yellow to Red
    this.triColors = this.genMultiHexArray([0x00ff00, 0xffff00, 0xff0000], 50)
  }

  /**
   * Break down hex value into RGB components
   *
   * @param input - hexadecimal IHexColorNumber or IHexColorString representing color
   * @returns R, G, B codes
   */
  private static getRGBComponents(
    input: IHexColorNumber | IHexColorString,
  ): IRGB {
    const hex = Halli.convertToNumber(input)
    const r = hex >> 16
    const g = (hex >> 8) & 0xff
    const b = hex & 0xff

    return {r, g, b, a: 255}
  }

  /**
   * Convert IHexColorNumber or IHexColorString to IHexColorNumber
   *
   * @param input - convert a hex code string to a number
   * @returns Number
   */
  protected static convertToNumber(
    input: IHexColorNumber | IHexColorString,
  ): IHexColorNumber {
    if (typeof input === 'string') {
      if (hexCheck.test(input)) {
        return Number(`0x${Halli.frontPad(input.replace('#', ''))}`)
      }
      return Number(input)
    }
    return input
  }

  /**
   * Useful for checking front padding of a hex color code
   *
   * @param hexStr - string representation of hex color code
   * @returns 0 front padded hex color code
   */
  private static frontPad(hexStr: string): IHexColorString {
    while (hexStr.length < 6) {
      hexStr = '0' + hexStr
    }
    return hexStr
  }

  /**
   * Select a color between color 1 (hc1) and color 2 (hc2)
   * based on an input ratio that denotes the point in the
   * gradient color should be selected from.
   *
   * @param hc1 - hexadecimal number representing color
   * @param hc2 - hexadecimal number representing color
   * @param ratio - the point in the gradient color to be selected
   * @returns IHexColorString color code in string form
   */
  public pickHex(
    hc1: IHexColorNumber | IHexColorString,
    hc2: IHexColorNumber | IHexColorString,
    ratio: number,
  ): IHexColorString {
    if (ratio > 1) ratio = 1

    // Break hc1 into RGB components
    const {r, g, b} = Halli.getRGBComponents(hc1)

    // RGB Delta between hc1 and hc2
    const {r: r2, g: g2, b: b2} = Halli.getRGBComponents(hc2)
    const rd = r2 - r
    const gd = g2 - g
    const bd = b2 - b

    // Calculate new color
    return `#${Halli.frontPad(
      (
        ((r + rd * ratio) << 16) |
        ((g + gd * ratio) << 8) |
        (b + bd * ratio)
      ).toString(16),
    )}`
  }

  /**
   * Select n=steps colors between color 1 (hc1) and color 2 (hc2)
   * and populate a return array.
   *
   * @param hc1 - hexadecimal number representing color
   * @param hc2 - hexadecimal number representing color
   * @param steps - the number of colors to generate between two colors
   * @returns array of string hex color codes
   */
  public genHexArray(
    hc1: IHexColorNumber | IHexColorString,
    hc2: IHexColorNumber | IHexColorString,
    steps: number,
  ): IHexColorString[] {
    const colorArray: IHexColorString[] = []

    // Place source color in as first element
    colorArray[0] = `#${Halli.frontPad(hc1.toString(16)).replace('#', '')}`

    // Break hc1 into RGB components
    const {r, g, b} = Halli.getRGBComponents(hc1)

    // RGB Delta between hc1 and hc2
    const {r: r2, g: g2, b: b2} = Halli.getRGBComponents(hc2)
    const rd = r2 - r
    const gd = g2 - g
    const bd = b2 - b

    // smoother gradient
    steps++

    // Determine new colors
    for (let i = 1; i < steps; i++) {
      // Where are we on the gradient?
      const ratio = i / steps

      // Calculate new color and add it to the array
      colorArray[i] = `#${Halli.frontPad(
        (
          ((r + rd * ratio) << 16) |
          ((g + gd * ratio) << 8) |
          (b + bd * ratio)
        ).toString(16),
      )}`
    }

    // Tack on the last color to complete the gradient
    colorArray.push(`#${Halli.frontPad(hc2.toString(16)).replace('#', '')}`)

    return colorArray
  }

  /**
   * genMultiHexArray will take an input array of any number of hex colors
   * and the number of steps to create a smooth gradient fade between the
   * colors in sequence.
   *
   * Example: input = [0x00FF00, 0xFFFF00, 0xFF0000] with steps = 10
   * will produce a Green to Yellow to Red gradient with 10 steps between
   * each source color. Total of 23 color codes.
   *
   * @param input - array of hexadecimal numbers representing colors
   * @param steps - the number of colors to generate between two colors
   * @returns array of string hex color codes
   */
  public genMultiHexArray(
    input: IHexColorNumber[] | IHexColorString[],
    steps: number,
  ): IHexColorString[] {
    const multiColor: IHexColorString[] = []

    // Find each sequential pair to compare
    for (let i = 0; i < input.length - 1; i++) {
      // Set hex colors
      const hc1 = input[i]
      const hc2 = input[i + 1]

      // Save first color
      multiColor.push(`#${Halli.frontPad(hc1.toString(16)).replace('#', '')}`)

      // Break hc1 into RGB components
      const {r, g, b} = Halli.getRGBComponents(hc1)

      // RGB Delta between hc1 and hc2
      const {r: r2, g: g2, b: b2} = Halli.getRGBComponents(hc2)
      const rd = r2 - r
      const gd = g2 - g
      const bd = b2 - b

      // Determine new colors
      for (let j = 1; j < steps; j++) {
        // Where are we on the gradient?
        const ratio = j / steps
        // Calculate new color and add it to the array
        multiColor.push(
          `#${Halli.frontPad(
            (
              ((r + rd * ratio) << 16) |
              ((g + gd * ratio) << 8) |
              (b + bd * ratio)
            ).toString(16),
          )}`,
        )
      }
    }

    // Add the last color to the end of the array.
    multiColor.push(`#${Halli.frontPad(input[input.length - 1].toString(16))}`)

    // Return the new array of colors.
    return multiColor
  }

  /**
   * Function will return an object with \{r,g,b,a\} set to appropriate values.
   *
   * @param hex - string representation of a color
   * @returns R, G, B, A codes
   */
  public static hexToRGBA(hex: IHexColorString): IRGB {
    // Break hc1 into RGB components
    const sanitized = hex.replace('#', '')
    return {
      // R
      r: parseInt(sanitized.substring(0, 2), 16),
      // G
      g: parseInt(sanitized.substring(2, 4), 16),
      // B
      b: parseInt(sanitized.substring(4, 6), 16),
      // Alpha
      a:
        sanitized.substring(6, 8) !== ''
          ? parseInt(sanitized.substring(6, 8), 16)
          : 255,
    }
  }

  /**
   * Function will return a 'rgba( r,g,b,a )' string set to the appropriate values.
   *
   * @param hex - string representation of a color
   * @returns rgba(r, g, b, a)
   */
  public static hexToRGBAstr(hex: string): IRGBACSSString {
    // Break hc1 into RGB components
    const rgba = Halli.hexToRGBA(hex)
    return `rgba(${rgba.r}, ${rgba.g}, ${rgba.b}, ${rgba.a})`
  }

  /**
   * Function will return a '#000000' string set to the appropriate values.
   *
   * @param hex - hexadecimal representation of a color
   * @returns `#000000`
   */
  public static convertToColorString(hex: number): string {
    // Break hc1 into RGB components
    return `#${hex.toString(16)}`
  }
}
